### DISCUSSION TITLE
Excessive fetching

### DISCUSSION THREAD

**Discussion by hyusetiawan** [gqty-dev/gqty#1605] [Q&A]
What's the lifecycle of a query with arguments? Ex: session?.chat({gid: props.gid}) 

Will it fetch when the react component state changes? Or only when props.gid gets a new value? 

Currently, if it's just property access like session?.use?. email it works as expected but when there's an argument, it seems to create a network request every time the react component renders

**Comment by vicary** [comment]
You may take a look at the [useQuery API](https://gqty.dev/api-reference/react/use-query), especially the options starts with refetch*

1. refetchInterval
2. refetchOnReconnect
3. refetchOnRender
4. refetchOnWindowVisible
5. retryPolicy

These are *soft refetches*, which ultimately checks your cache validity (`maxAge` and `staleWhileRevalidate`), and only fetches when it encounters a stale, expired or missing cache value.

1. `maxAge` determines cache expiry
2. `staleWhileRevalidate` determines staleness

**Comment by hyusetiawan** [comment]
I saw them but it's still unclear to me when it actually triggers an actual request, because it triggers a request repeatedly for me. What's considered to be stale?

**Reply by vicary** [reply]
The behavior depends on the `cachePolicy` option. I will describe the *default* mode below, which is the most complicated one. Other cache modes just let you skip a specific part of the following process.

Cache values experience 3 stages before they are gone.

1. Fresh: The value is fetched from network and stored in the cache for shorter than `maxAge`. Render gives you the value, no fetch will happen.
2. Stale: The stored cache value has passed `maxAge`, but shorter than `maxAge + staleWhileRevalidate`. Renders still gives you the data, but it will refetch in the background to keep things fresh.
3. Expired: The cache value passed `maxAge + staleWhileRevalidate`, cache values are now open up for V8 to remove it. Normally you cant see these values and they must be refetched, but you can use `force-cache` to try a best-effort rendering for e.g. when cell network is temporarily lost.


With a freshly generated client, you have maxAge 0 and SWR in 5 secs, so you should see frequent refetches. This is done in favor of rapid iteration when you just start a fresh project, don't bother with network optimization until the basic works are done.

**Reply by hyusetiawan** [reply]
The frequent refetch makes it a rather bad default no? Is it possible to mimic a typical graphql client refetch? So at least the dev ux expectation still matches with other graphql clients (urql/Apollo/etc) while maintaining a simple api, which is the framework's main value proposition?

**Reply by vicary** [reply]
We introduced quite a few popular ideas into GQty, you will feel the refetches familiar if you have used React Query before.

Making it feels more like Apollo is easy, just do the following:
1. `maxAge: Infinity`
2. `query.$refetch(true)`

**Reply by hyusetiawan** [reply]
Gotcha, when do you call the  $refetch in this case?

**Reply by vicary** [reply]
When you give `maxAge` the `Infinity` treatment, it's up to you to call `$refetch` when needed.

It could be a reload button in a data table, a pull-down trigger from an infinite loading list... etc.

**Reply by hyusetiawan** [reply]
it seems maxAge is not a property on useQuery if my research is right, it's in the cache?
I am not sure how I can integrate cache to my queries, the example says `new Cache(...` does it automagically get read by the `useQuery` hook?

UPDATE:
i found it in the index.ts

**Reply by hyusetiawan** [reply]
@vicary  apologies for this barrage of questions for the last couple of days, feel like we are close to getting it to a productive state if you could just bear with us 🙏
I am still having issues with excessive refetching, your suggestions did help by changing the `refetchOnRender` to false and `maxAge` to infinity, but now that caused the query to not fetch at all after given a new id to fetch.
index.ts cache is:
```
const cache = new Cache(
  undefined,
  /**
   * Default option is immediate cache expiry but keep it for 5 minutes,
   * allowing soft refetches in background.
   */
  {
    maxAge: Infinity,
    staleWhileRevalidate: 5 * 60 * 1000,
    normalization: true,
  },
)
```
and here a snippet of the component
```
export function ManageAgentForm(props: Props) {
  const { session } = useQuery({
    refetchOnRender: false,
    refetchOnWindowVisible: false,
  })
  const [deleteAgent, deleteAgentState] = useMutation('deleteAgent', (isDeleted) => {
    isDeleted && props.onDelete()
  })
  const agent = session?.user.agent({ gid: props.gid })
  console.log('AGENT: ', agent?.name)
```
if `props.gid` changes, the query won't refetch, I guess this falls under the `refetchOnRender` flag, but if I turn that to `true` the component will keep re-rendering and refetching (actually sending requests infinitely). 

This video shows, when `refetchOnRender` is false, on first load, it will fetch the data then i click around, the new data given the new id won't be fetched (shows undefined), changing to `refetchOnRender` to true when I click an id that is not cached, it will trigger infinite fetching.

The only way I could make it work is to do the following:
```
  useEffect(() => {
    $refetch()
  }, [props.gid]) // cannot add $refetch to the dependency list as it will cause infinite refetch
```
given the workarounds I have to do to make this common use case of "passing new id, fetch", I am sure I am missing something rather simple in my setup.

https://github.com/gqty-dev/gqty/assets/1062920/8ff0b833-f749-446a-bba4-df697f1d9f63

**Reply by vicary** [reply]
It's unclear what exactly is being fetched, would you please share a minimum reproducing repo?
